Controller.cpp

```diff
@@ -443,22 +357,170 @@ namespace {
+		using namespace std::this_thread;
+		using namespace std::chrono;
+
+		if (dev == nullptr)
+			throw ControllerException("Unable to open controller device: dev was nullptr.");
+		if (dev->product_id != Procon_ID)
+			throw ControllerException("Unable to open controller device: product id was not a Switch Pro Controller.");
+		if (wcscmp(dev->serial_number, L"000000000001") != 0)
+			throw ControllerException("Device is not connected via USB");
+
+		device.reset(hid_open_path(dev->path));
+		if (!device)
+			throw ControllerException("Unable to open controller device: device path could not be opened.");
+
+		// Start handshake
+		if (!exchange(handshake)) {
+			throw ControllerException("Handshake failed.");
+		}
+		exchange(switchBaudrate);
+		exchange(handshake);
+		exchange(HIDOnlyMode);
+
+		//sendSubcommand(0x1, rumbleCommand, enable);
+		// Set report mode to 0x30 (we now get the same data as in the bluetooth case)
+		std::array<uchar, 2> report{ 0x30, 0x00 };
+		sendSubcommand(0x01, 0x03, report);
+		sendSubcommand(0x01, imuDataCommand, enable);
+		sendSubcommand(0x01, ledCommand, led);
+
+		// Read calibration data (use factory calibration for now)
+		readAnalogStickCalibrationData();
+		readMotionSensorCalibrationData();
+
+		if (XOutputPlugIn(port) != ERROR_SUCCESS) {
+			device.reset(nullptr);
+			throw ControllerException("Unable to plugin XOutput controller.");
+		}
+		_connected = true;
+		sleep_for(milliseconds(100));
+		//updateStatus();
+	}
+
+	void Controller::readAnalogStickCalibrationData()
+	{
+		auto factory_stick_cal = getSpiData(0x603D, 0x12).value();
+
+		calib.leftCenter.x = (factory_stick_cal[4] << 8) & 0xF00 | factory_stick_cal[3];
+		calib.leftCenter.y = (factory_stick_cal[5] << 4) | (factory_stick_cal[4] >> 4);
+		calib.left.x.min = calib.leftCenter.x - ((factory_stick_cal[7] << 8) & 0xF00 | factory_stick_cal[6]);
+		calib.left.x.max = calib.leftCenter.x + ((factory_stick_cal[1] << 8) & 0xF00 | factory_stick_cal[0]);
+		calib.left.y.min = calib.leftCenter.y - ((factory_stick_cal[8] << 4) | (factory_stick_cal[7] >> 4));
+		calib.left.y.max = calib.leftCenter.y + ((factory_stick_cal[2] << 4) | (factory_stick_cal[1] >> 4));
+
+		calib.rightCenter.x = (factory_stick_cal[10] << 8) & 0xF00 | factory_stick_cal[9];
+		calib.rightCenter.y = (factory_stick_cal[11] << 4) | (factory_stick_cal[10] >> 4);
+		calib.right.x.min = calib.rightCenter.x - ((factory_stick_cal[13] << 8) & 0xF00 | factory_stick_cal[12]);
+		calib.right.x.max = calib.rightCenter.x + ((factory_stick_cal[16] << 8) & 0xF00 | factory_stick_cal[15]);
+		calib.right.y.min = calib.rightCenter.y - ((factory_stick_cal[14] << 4) | (factory_stick_cal[13] >> 4));
+		calib.right.y.max = calib.rightCenter.y + ((factory_stick_cal[17] << 4) | (factory_stick_cal[16] >> 4));
+	}
+
+	void Controller::readMotionSensorCalibrationData()
+	{
+		auto factory_sensor_cal = getSpiData(0x6020, 0x18).value();
+		auto uint16ToInt16 = [](uint16_t a) {return *(int16_t*)(&a); };
```

Comment:
Use C++ casts.

```c++
auto uint16ToInt16 = [](uint16_t a) {return *reinterpret_cast<int16_t*>(&a); };
```

Reply:
Yes, I know. But I first wanted this to work somehow and start a discussion about the whole code. I will change this later.